li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -299,6 +324,55 @@ public void abortTransaction() throws ProducerFencedException {
     throw new UnsupportedOperationException("Not supported");
   }
 
+  public LiKafkaProducerConfig getCommonProducerConfigs() {
+    return _commonProducerConfigs;
+  }
+
+  public void reloadConfig(Map<String, String> newConfigs, UUID commandId) {
+    // Go over each producer, flush and close. Since each per-cluster producer will be instantiated when the client began
+    // producing to that cluster, we just need to clear the mappings and update the configs
+    // since this method is invoked by marioClient, it should be non-blocking, so we create another thread doing above
+    LOG.info("Starting reloadConfig for LiKafkaProducers in clusterGroup {} with new configs {}", _clusterGroup, newConfigs);
+    Thread t = new Thread(() -> {
+        flushAndCloseExistingProducers(newConfigs, commandId);
+    });
+    t.setDaemon(true);
+    t.setName("LiKafkaProducer-reloadConfig-" + _clusterGroup.getName());
+    t.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+      public void uncaughtException(Thread t, Throwable e) {
+        throw new KafkaException("Thread " + t.getName() + " throws exception", e);
+      }
+    });
+
+    t.start();
+  }
+
+  // Do synchronization on the method that will be flushing and closing the producers since it will be handled by a different thread
+  synchronized void flushAndCloseExistingProducers(Map<String, String> newConfigs, UUID commandId) {
+    ensureOpen();
+
+    flushNoLock(RELOAD_CONFIG_EXECUTION_TIME_OUT.toMillis(), TimeUnit.MILLISECONDS);
+    closeNoLock(RELOAD_CONFIG_EXECUTION_TIME_OUT.toMillis(), TimeUnit.MILLISECONDS);
+
+    // update existing configs with newConfigs
+    Map<String, Object> configMap = _commonProducerConfigs.originals();
+    configMap.putAll(newConfigs);
+    _commonProducerConfigs = new LiKafkaProducerConfig(configMap);
+
+    // reset the state of _producers and _closed so that subsequent send will try to instantiate the producers with new configs
+    _producers.clear();
+    _closed = false;
```

Comment:
why do we need to mess with _closed

Reply:
I think it's just to avoid NPE or other weird errors when users call other methods after producers are already closed.